sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -1057,7 +1057,7 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
     val pythonEvals = collect(joinNode.get) {
       case p: BatchEvalPythonExec => p
     }
-    assert(pythonEvals.size == 2)
+    assert(pythonEvals.size == 4)
```

Comment:
This looks like a big problem. Can you investigate why we run python UDF 2 more times?

Reply:
It is because we will infer two `isnotnull(cast(pythonUDF0 as int))`:
```
== Optimized Logical Plan ==
Project [a#225, b#226, c#236, d#237]
+- Join Inner, (CAST(udf(cast(a as string)) AS INT)#250 = CAST(udf(cast(c as string)) AS INT)#251)
   :- Project [_1#220 AS a#225, _2#221 AS b#226, cast(pythonUDF0#253 as int) AS CAST(udf(cast(a as string)) AS INT)#250]
   :  +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#253]
   :     +- Project [_1#220, _2#221]
   :        +- Filter isnotnull(cast(pythonUDF0#252 as int))
   :           +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#252]
   :              +- LocalRelation [_1#220, _2#221]
   +- Project [_1#231 AS c#236, _2#232 AS d#237, cast(pythonUDF0#255 as int) AS CAST(udf(cast(c as string)) AS INT)#251]
      +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#255]
         +- Project [_1#231, _2#232]
            +- Filter isnotnull(cast(pythonUDF0#254 as int))
               +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#254]
                  +- LocalRelation [_1#231, _2#232]
```

Reply:
OK, so it's because filter push down can lead to extra expression evaluation?

Reply:
Yes. If we disable `spark.sql.constraintPropagation.enabled`, the plan is:
```
== Optimized Logical Plan ==
Project [a#225, b#226, c#236, d#237]
+- Join Inner, (CAST(udf(cast(a as string)) AS INT)#250 = CAST(udf(cast(c as string)) AS INT)#251)
   :- Project [_1#220 AS a#225, _2#221 AS b#226, cast(pythonUDF0#252 as int) AS CAST(udf(cast(a as string)) AS INT)#250]
   :  +- BatchEvalPython [udf(cast(_1#220 as string))], [pythonUDF0#252]
   :     +- LocalRelation [_1#220, _2#221]
   +- Project [_1#231 AS c#236, _2#232 AS d#237, cast(pythonUDF0#253 as int) AS CAST(udf(cast(c as string)) AS INT)#251]
      +- BatchEvalPython [udf(cast(_1#231 as string))], [pythonUDF0#253]
         +- LocalRelation [_1#231, _2#232]
```

Reply:
@wangyum @cloud-fan Seems we can just put the rule after `InferFiltersFromConstraints`, and actually after the `earlyScanPushDownRules`. The python UDF test passed in my pr, see https://github.com/apache/spark/pull/36874.

Reply:
I think pull out has 3 advantages:
1. Reduce complex join key runs from 3 to 2 for SMJ.
2. Infer additional filters, sometimes can avoid data skew. For example: https://github.com/apache/spark/pull/28642
3. Avoid other rules also handle this case. For example: https://github.com/apache/spark/blob/dee7396204e2f6e7346e220867953fc74cd4253d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushPartialAggregationThroughJoin.scala#L325-L327

It has two disadvantage:
1. Increase complex join key runs from 1 to 2 for BHJ.
2. It may increase the data size of shuffle. For example: the join key is: `concat(col1, col2, col3, col4 ...)`.

Personally, I think this rule is valuable. We have been using this rule for half a year.

Reply:
> Increase complex join key runs from 1 to 2 for BHJ.

We can check if the poll out side can be broadcast so it should not be a blocker ?

> It may increase the data size of shuffle. For example: the join key is: concat(col1, col2, col3, col4 ...).

This is really a trade-off, one conservative option may be: We only poll out the complex keys which the inside attribute is not the final output. So we can avoid the extra shuffle data as far as possible, for example:
```sql
SELECT a FROM t1 JOIN t2 on t1.a = t2.x + 1;
```

And a config should be introduced for enable or disable easily.